{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: lifelines in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (0.28.0)\n",
      "Requirement already satisfied: numpy<2.0,>=1.14.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from lifelines) (1.26.4)\n",
      "Requirement already satisfied: scipy>=1.2.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from lifelines) (1.12.0)\n",
      "Requirement already satisfied: pandas>=1.2.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from lifelines) (2.2.0)\n",
      "Requirement already satisfied: matplotlib>=3.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from lifelines) (3.8.3)\n",
      "Requirement already satisfied: autograd>=1.5 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from lifelines) (1.6.2)\n",
      "Requirement already satisfied: autograd-gamma>=0.3 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from lifelines) (0.5.0)\n",
      "Requirement already satisfied: formulaic>=0.2.2 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from lifelines) (1.0.1)\n",
      "Requirement already satisfied: future>=0.15.2 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from autograd>=1.5->lifelines) (1.0.0)\n",
      "Requirement already satisfied: interface-meta>=1.2.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from formulaic>=0.2.2->lifelines) (1.3.0)\n",
      "Requirement already satisfied: typing-extensions>=4.2.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from formulaic>=0.2.2->lifelines) (4.9.0)\n",
      "Requirement already satisfied: wrapt>=1.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from formulaic>=0.2.2->lifelines) (1.16.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib>=3.0->lifelines) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib>=3.0->lifelines) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib>=3.0->lifelines) (4.49.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib>=3.0->lifelines) (1.4.5)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib>=3.0->lifelines) (23.2)\n",
      "Requirement already satisfied: pillow>=8 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib>=3.0->lifelines) (10.2.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib>=3.0->lifelines) (3.1.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib>=3.0->lifelines) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from pandas>=1.2.0->lifelines) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from pandas>=1.2.0->lifelines) (2024.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib>=3.0->lifelines) (1.16.0)\n",
      "Requirement already satisfied: Pgenlib in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (0.90.1)\n",
      "Requirement already satisfied: numpy in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (1.26.4)\n",
      "Collecting sci_palettes\n",
      "  Using cached sci_palettes-1.0.1-py3-none-any.whl (7.8 kB)\n",
      "Collecting seaborn (from sci_palettes)\n",
      "  Using cached seaborn-0.13.2-py3-none-any.whl.metadata (5.4 kB)\n",
      "Requirement already satisfied: matplotlib in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from sci_palettes) (3.8.3)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->sci_palettes) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->sci_palettes) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->sci_palettes) (4.49.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->sci_palettes) (1.4.5)\n",
      "Requirement already satisfied: numpy<2,>=1.21 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->sci_palettes) (1.26.4)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->sci_palettes) (23.2)\n",
      "Requirement already satisfied: pillow>=8 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->sci_palettes) (10.2.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->sci_palettes) (3.1.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->sci_palettes) (2.8.2)\n",
      "Requirement already satisfied: pandas>=1.2 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from seaborn->sci_palettes) (2.2.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from pandas>=1.2->seaborn->sci_palettes) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from pandas>=1.2->seaborn->sci_palettes) (2024.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib->sci_palettes) (1.16.0)\n",
      "Using cached seaborn-0.13.2-py3-none-any.whl (294 kB)\n",
      "Installing collected packages: seaborn, sci_palettes\n",
      "Successfully installed sci_palettes-1.0.1 seaborn-0.13.2\n",
      "Collecting SciencePlots\n",
      "  Using cached SciencePlots-2.1.1-py3-none-any.whl.metadata (11 kB)\n",
      "Requirement already satisfied: matplotlib in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from SciencePlots) (3.8.3)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->SciencePlots) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->SciencePlots) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->SciencePlots) (4.49.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->SciencePlots) (1.4.5)\n",
      "Requirement already satisfied: numpy<2,>=1.21 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->SciencePlots) (1.26.4)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->SciencePlots) (23.2)\n",
      "Requirement already satisfied: pillow>=8 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->SciencePlots) (10.2.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->SciencePlots) (3.1.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from matplotlib->SciencePlots) (2.8.2)\n",
      "Requirement already satisfied: six>=1.5 in /home/xutingfeng/miniforge3/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib->SciencePlots) (1.16.0)\n",
      "Using cached SciencePlots-2.1.1-py3-none-any.whl (16 kB)\n",
      "Installing collected packages: SciencePlots\n",
      "Successfully installed SciencePlots-2.1.1\n"
     ]
    }
   ],
   "source": [
    "!pip install lifelines\n",
    "!pip install Pgenlib\n",
    "!pip install numpy \n",
    "!pip install sci_palettes\n",
    "!pip install SciencePlots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pgenlib as pg\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pgen file have rows (ind):  467491\n",
      "pgen file have columns (snp) :  427\n",
      "pgen file is hard call:  False\n"
     ]
    }
   ],
   "source": [
    "test_pvar_path = \"test/ANGPTL3.pvar\"\n",
    "test_pgen_path = \"test/ANGPTL3.pgen\"\n",
    "\n",
    "pfile = pg.PgenReader(bytes(test_pgen_path, \"utf-8\"))\n",
    "print(\"pgen file have rows (ind): \", pfile.get_raw_sample_ct())\n",
    "print(\"pgen file have columns (snp) : \", pfile.get_variant_ct())\n",
    "print(\"pgen file is hard call: \", pfile.hardcall_phase_present())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pvar variants counts: 427\n",
      "pvar variant 1: 1:62597543:T:C\n"
     ]
    }
   ],
   "source": [
    "pvar = pg.PvarReader(bytes(test_pvar_path, \"utf-8\"))\n",
    "print(\"pvar variants counts:\", pvar.get_variant_ct())\n",
    "print(\"pvar variant 1:\", pvar.get_variant_id(0).decode())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`PgenReader`是最为主要的对象\n",
    "\n",
    "```python\n",
    "class PgenReader:\n",
    "    \"\"\"\n",
    "    .pgen 或 .bed 文件的读取器。\n",
    "\n",
    "    Parameters:\n",
    "        filename (bytes): 文件名，如果文件不存在或无效，则会引发异常。\n",
    "        raw_sample_ct (int, optional): .bed 文件所需的原始样本数，如果不提供将引发异常；对于 .pgen 文件，可选。如果为 .pgen 文件提供了该参数，则如果样本数不匹配 .pgen 中的值，则会引发异常。\n",
    "        variant_ct (int, optional): 变异位点数，始终可选。如果不提供，则会检查 .bed 文件的大小或 .pgen 中显式存储的值，如果不匹配则引发异常。\n",
    "        sample_subset (np.ndarray[np.uint32], optional): （从零开始的）索引数组，告诉读取器仅在加载或计算基因型时考虑指定的样本。如果不提供，则表示包括所有样本。如果提供了值，则会检查数组中的值是否在0到(raw_sample_ct - 1)的范围内，并且是否严格递增，如果不是则引发异常。\n",
    "        allele_idx_offsets (np.ndarray[np.uintp], optional): 多等位基因型的偏移量数组。当提供时，应该是一个长度为 (variant_ct+1) 的 np.uintp 数组，其中 allele_idx_offsets[i+1] - allele_idx_offsets[i] 是 0-based 变异位点 i 的等位基因数目。\n",
    "        pvar (object, optional): 如果提供了 pvar，则从中提取 variant_ct 和 allele_idx_offsets。（不能同时提供 pvar 和 variant_ct/allele_idx_offsets）\n",
    "\n",
    "    Attributes:\n",
    "        filename (bytes): 文件名。\n",
    "        raw_sample_ct (int): .bed 文件所需的原始样本数。\n",
    "        variant_ct (int): 变异位点数。\n",
    "        sample_subset (np.ndarray[np.uint32]): 样本子集索引数组。\n",
    "        allele_idx_offsets (np.ndarray[np.uintp]): 多等位基因型的偏移量数组。\n",
    "\n",
    "    Raises:\n",
    "        ValueError: 如果文件不存在、无效或参数不匹配要求时引发异常。\n",
    "    \"\"\"\n",
    "    def __init__(self, filename: bytes, raw_sample_ct: int = None, variant_ct: int = None, \n",
    "                 sample_subset: np.ndarray[np.uint32] = None, allele_idx_offsets: np.ndarray[np.uintp] = None, \n",
    "                 pvar: object = None) -> None:\n",
    "        pass  # 初始化函数，打开 .pgen 或 .bed 文件\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pgen file have rows (ind):  467491\n",
      "pgen file have columns (snp) :  427\n",
      "pgen file is hard call:  False\n"
     ]
    }
   ],
   "source": [
    "test_pvar_path = 'pgen/ANGPTL3.pvar'\n",
    "test_pgen_path = 'pgen/ANGPTL3.pgen'\n",
    "\n",
    "pfile = pg.PgenReader(bytes(test_pgen_path, 'utf-8'))\n",
    "print(\"pgen file have rows (ind): \", pfile.get_raw_sample_ct())\n",
    "print(\"pgen file have columns (snp) : \", pfile.get_variant_ct())\n",
    "print(\"pgen file is hard call: \", pfile.hardcall_phase_present())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'pgenlib.PgenReader' object has no attribute 'get_variant_id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m pfile \u001b[38;5;241m=\u001b[39m pg\u001b[38;5;241m.\u001b[39mPgenReader(\u001b[38;5;28mbytes\u001b[39m(test_pgen_path, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mutf-8\u001b[39m\u001b[38;5;124m'\u001b[39m), pvar \u001b[38;5;241m=\u001b[39m pg\u001b[38;5;241m.\u001b[39mPvarReader(\u001b[38;5;28mbytes\u001b[39m(test_pvar_path, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mutf-8\u001b[39m\u001b[38;5;124m'\u001b[39m)))\n\u001b[0;32m----> 3\u001b[0m \u001b[43mpfile\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_variant_id\u001b[49m(\u001b[38;5;241m32\u001b[39m)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'pgenlib.PgenReader' object has no attribute 'get_variant_id'"
     ]
    }
   ],
   "source": [
    "pfile = pg.PgenReader(bytes(test_pgen_path, 'utf-8'), pvar = pg.PvarReader(bytes(test_pvar_path, 'utf-8')))\n",
    "\n",
    "pfile.get_variant_id(32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, ..., 0, 0, 0], dtype=int8)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- read function\n",
    "\n",
    "read(int32_t variant_idx, np.ndarray[np.int{8,32,64}_t] geno_int_out,\n",
    "       uint32_t allele_idx = 1)\n",
    "- variant_idx:int32, 接受索引\n",
    "- geno_int_out: np.array(dtype=np.int{8,32,64}) 用来作为buf，接受输出，最终的输出结果存放在这个变量里面\n",
    "- allele_idx: int32, default 1 ，1 返回的计数是基于alt的，0则是基于REF\n",
    "\n",
    "```python\n",
    "def read(variant_idx: int, geno_int_out: np.ndarray[np.int8 | np.int32 | np.int64], allele_idx: int = 1) -> None:\n",
    "    \"\"\"\n",
    "    读取指定索引的变异位点信息。\n",
    "\n",
    "    Parameters:\n",
    "        variant_idx (int): 变异位点的索引，使用 int32 类型表示。\n",
    "        geno_int_out (np.ndarray): 用于存储基因型信息的 numpy 数组，dtype 为 np.int8、np.int32 或 np.int64。\n",
    "        allele_idx (int, optional): 等位基因索引，默认为 1。当 allele_idx 为 1 时，返回的基因型信息是基于替代等位基因的计数；当 allele_idx 为 0 时，返回的基因型信息是基于参考等位基因的计数。\n",
    "\n",
    "    Returns:\n",
    "        None: 函数没有返回值，但是会将结果存储在 geno_int_out 中。\n",
    "    \"\"\"\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, ..., 0, 0, 0], dtype=int8)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 读取一个人的全部SNP\n",
    "buf = np.empty(pfile.get_raw_sample_ct(), np.int8)  # why use this ? - Author tried the interface without the required outbuffer, and it could take twice as long due to all the extra allocation/deallocation. \n",
    "pfile.read(0, buf) # 返回结果存放在buf里面\n",
    "buf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "def read_dosages(variant_idx: int, floatarr_out: np.ndarray[np.float32 | np.float64], allele_idx: int = 1) -> None:\n",
    "    \"\"\"\n",
    "    读取指定索引的变异位点剂量信息。\n",
    "\n",
    "    Parameters:\n",
    "        variant_idx (int): 变异位点的索引，使用 uint32_t 类型表示。\n",
    "        floatarr_out (np.ndarray): 用于存储剂量信息的 numpy 数组，dtype 为 np.float32 或 np.float64。\n",
    "        allele_idx (int, optional): 等位基因索引，默认为 1。当 allele_idx 为 1 时，返回的剂量信息是基于替代等位基因的剂量；当 allele_idx 为 0 时，返回的剂量信息是基于参考等位基因的剂量。\n",
    "\n",
    "    Returns:\n",
    "        None: 函数没有返回值，但是会将结果存储在 floatarr_out 中。\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0., 0., 0., ..., 0., 0., 0.], dtype=float32)"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "buf = np.empty(pfile.get_raw_sample_ct(), np.float32) \n",
    "pfile.read_dosages(0, buf)\n",
    "buf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "def read_range(variant_idx_start: int, variant_idx_end: int,\n",
    "               geno_int_out: np.ndarray[np.int8 | np.int32 | np.int64, mode='c', ndim=2],\n",
    "               allele_idx: int = 1, sample_maj: int = 0) -> None:\n",
    "    \"\"\"\n",
    "    读取指定范围内的变异位点信息。\n",
    "\n",
    "    Parameters:\n",
    "        variant_idx_start (int): 变异位点的起始索引，使用 uint32_t 类型表示。\n",
    "        variant_idx_end (int): 变异位点的结束索引，使用 uint32_t 类型表示。\n",
    "        geno_int_out (np.ndarray): 用于存储基因型信息的二维 numpy 数组，dtype 为 np.int8、np.int32 或 np.int64。mode 为 \"c\"，ndim 为 2\n",
    "        allele_idx (int, optional): 等位基因索引，默认为 1。当 allele_idx 为 1 时，返回的基因型信息是基于替代等位基因的计数；当 allele_idx 为 0 时，返回的基因型信息是基于参考等位基因的计数。\n",
    "        sample_maj (int, optional): 不知道功能，没有文档解释\n",
    "\n",
    "    Returns:\n",
    "        None: 函数没有返回值，但是会将结果存储在 geno_int_out 中。\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10, 467491)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[0, 0, 0, ..., 0, 0, 0],\n",
       "       [0, 0, 0, ..., 0, 0, 0],\n",
       "       [0, 0, 0, ..., 0, 0, 0],\n",
       "       ...,\n",
       "       [0, 0, 0, ..., 0, 0, 0],\n",
       "       [0, 0, 0, ..., 0, 0, 0],\n",
       "       [0, 0, 0, ..., 0, 0, 0]], dtype=int8)"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start =0 \n",
    "end = 10 \n",
    "nums = end - start\n",
    "\n",
    "buf = np.empty(pfile.get_raw_sample_ct(), np.int8).reshape(1, -1).repeat(nums, axis=0) # shape (num, sample)\n",
    "print(buf.shape)\n",
    "pfile.read_range(start, end, buf)\n",
    "buf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "def read_list(variant_idxs: np.ndarray[np.uint32], \n",
    "              geno_int_out: np.ndarray[np.int8 | np.int32 | np.int64, mode='c', ndim=2], \n",
    "              allele_idx: int = 1, sample_maj: int = 0) -> None:\n",
    "    \"\"\"\n",
    "    读取指定索引列表中的变异位点信息。\n",
    "\n",
    "    Parameters:\n",
    "        variant_idxs (np.ndarray): 变异位点索引的数组，dtype 为 np.uint32。\n",
    "        geno_int_out (np.ndarray): 用于存储基因型信息的二维 numpy 数组，dtype 为 np.int8、np.int32 或 np.int64。mode 为 \"c\"，ndim 为 2。\n",
    "        allele_idx (int, optional): 等位基因索引，默认为 1。当 allele_idx 为 1 时，返回的基因型信息是基于替代等位基因的计数；当 allele_idx 为 0 时，返回的基因型信息是基于参考等位基因的计数。\n",
    "        sample_maj (int, optional): 样本主要等位基因的索引，默认为 0。\n",
    "\n",
    "    Returns:\n",
    "        None: 函数没有返回值，但是会将结果存储在 geno_int_out 中。\n",
    "    \"\"\"\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 包裹代码\n",
    "```python\n",
    "from typing import Union, List, overload\n",
    "# from \n",
    "\n",
    "class PgenReaderFull():\n",
    "    def __init__(self, pfile_path:str, pgen_path:str =None, pvar_path:str=None, sample_path:str=None ):\n",
    "\n",
    "        self.pgen_path:str = pgen_path if pgen_path is not None else pfile_path + '.pgen'\n",
    "        self.pvar_path:str = pvar_path if pvar_path is not None else pfile_path + '.pvar'\n",
    "        self.sample_path:str = sample_path if sample_path is not None else pfile_path + '.psam'\n",
    "\n",
    "        self.pvar:pg.PvarReader = pg.PvarReader(bytes(self.pvar_path, 'utf-8'))\n",
    "        self._init_var() # this may be too large to load\n",
    "        \n",
    "        self.sample:pd.DataFrame = pd.read_csv(self.sample_path, sep='\\t')\n",
    "        self._init_sample()\n",
    "\n",
    "        self.pgen:pg.PgenReader = pg.PgenReader(bytes(self.pgen_path, 'utf-8'), pvar=self.pvar, \n",
    "                                                sample_subset=np.arange(len(self.sample_list), dtype=np.uint32)\n",
    "                                                )\n",
    "\n",
    "\n",
    "        # basic information\n",
    "        self.sample_ct:int = self.get_raw_sample_ct() # number of samples as same as get_raw_sample_ct, but only call once\n",
    "        self.variant_ct:int = self.get_variant_ct() # number of variants as same as get_variant_ct, but only call once\n",
    "\n",
    "\n",
    "    @overload\n",
    "    def get_variant_ids(self, variant_idx:List[int])->List[str]:...\n",
    "    @overload\n",
    "    def get_variant_ids(self, variant_idx:int)->str:...\n",
    "\n",
    "    def get_variant_ids(self, variant_idx:Union[int, List[int]])->Union[str, List[str]]:\n",
    "        \"\"\"\n",
    "        from PvarReader.get_variant_id \n",
    "        Modified version for interface        \n",
    "        \"\"\"\n",
    "        if isinstance(variant_idx, int):\n",
    "            return self.pvar.get_variant_id(variant_idx).decode('utf-8')\n",
    "        else:\n",
    "            return [self.get_variant_ids(i) for i in variant_idx]\n",
    "        \n",
    "    @overload\n",
    "    def get_variant_idx(self, variant_id:str)->int:...\n",
    "    @overload\n",
    "    def get_variant_idx(self, variant_id:List[str])->List[int]:...\n",
    "    def get_variant_idx(self, variant_id:Union[str, List[str]])->Union[int, List[int]]:\n",
    "        \"\"\"\n",
    "        Modified version for interface\n",
    "        \"\"\"\n",
    "        if isinstance(variant_id, str):\n",
    "            return self.var_list.index(variant_id)\n",
    "        else:\n",
    "            return [self.get_variant_idx(i) for i in variant_id]\n",
    "    def get_sample_ids(self, sample_idx:Union[int, List[int]])->Union[str, List[str]]:\n",
    "        \"\"\"\n",
    "        TF version for get sample ids\n",
    "        \"\"\"\n",
    "        if isinstance(sample_idx, int):\n",
    "            return self.sample_list[sample_idx]\n",
    "        else:\n",
    "            return [self.get_sample_ids(i) for i in sample_idx]\n",
    "\n",
    "\n",
    "    def _init_sample(self):\n",
    "        \"\"\"\n",
    "        load from psam and to list[[FID, IID]]\n",
    "        sample_list = [[FID, IID], ...]\n",
    "        \"\"\"\n",
    "        self.sample_list = pd.read_csv(self.sample_path, sep='\\t').iloc[:, :2].values.tolist()\n",
    "    def _init_var(self):\n",
    "        \"\"\"\n",
    "        load from pvar and to list[variant_id]\n",
    "        \"\"\"\n",
    "        self.var_list = pd.read_csv(self.pvar_path, sep='\\t').iloc[:, 2].tolist()\n",
    "\n",
    "    def get_raw_sample_ct(self) -> int:\n",
    "        \"\"\"\n",
    "        from PgenReader.get_raw_sample_ct\n",
    "        \"\"\"\n",
    "        return self.pgen.get_raw_sample_ct()\n",
    "\n",
    "    def get_variant_ct(self) -> int:\n",
    "        \"\"\"\n",
    "        from PgenReader.get_variant_ct\n",
    "        \"\"\"\n",
    "        return self.pgen.get_variant_ct()\n",
    "    \n",
    "    def hardcall_phase_present(self) -> bool:\n",
    "        \"\"\"\n",
    "        from PgenReader.hardcall_phase_present\n",
    "        \"\"\"\n",
    "        return self.pgen.hardcall_phase_present()\n",
    "    def read(self, variant_idx, sample_idx=None, allele_idx=1)->np.ndarray:\n",
    "        \"\"\"\n",
    "        from PgenReader.read with modified \n",
    "\n",
    "        Parameters\n",
    "            - variant_idx:int32, index of the variant to read\n",
    "            - allele_idx (int, optional): 等位基因索引，默认为 1。当 allele_idx 为 1 时，返回的基因型信息是基于ALT等位基因的计数；当 allele_idx 为 0 时，返回的基因型信息是基于REF等位基因的计数。\n",
    "\n",
    "        \"\"\"\n",
    "        \n",
    "        sample_idx = np.array(sample_idx) if sample_idx is not None else np.array(range(self.sample_ct))\n",
    "\n",
    "        # use buf to store the result; the author of pgenlib think this is faster \n",
    "        buf = np.empty(self.sample_ct, dtype=np.int8)\n",
    "        self.pgen.read(variant_idx = variant_idx, \n",
    "                       geno_int_out = buf,\n",
    "                       allele_idx = allele_idx,\n",
    "                       \n",
    "                       )\n",
    "        return buf[sample_idx]\n",
    "    def read_range(self, variant_idx_start, variant_idx_end, allele_idx=1, sample_idx=None)->np.ndarray:\n",
    "        \"\"\"\n",
    "        from PgenReader.read_range\n",
    "        Parameters\n",
    "            - variant_idx_start: int32, index of the first variant to read\n",
    "            - variant_idx_end: int32, index of the last variant to read\n",
    "            - allele_idx: int32, index of the allele to read, defualt 1 for ALT count, while 0 for REF count\n",
    "            - sample_idx: int32, index of the sample to read\n",
    "        \"\"\"\n",
    "        sample_idx = np.array(sample_idx) if sample_idx is not None else np.array(range(self.sample_ct))\n",
    "\n",
    "        buf = np.empty((self.sample_ct, variant_idx_end-variant_idx_start), dtype=np.int8)\n",
    "        self.pgen.read_range(variant_idx_start = variant_idx_start, \n",
    "                             variant_idx_end = variant_idx_end,\n",
    "                             geno_int_out = buf,\n",
    "                             allele_idx = allele_idx,\n",
    "                             )\n",
    "        \n",
    "        return buf[:, sample_idx]\n",
    "    def read_list(self, variant_idxs:list,  allele_idx=1, sample_idx=None) -> np.ndarray:\n",
    "        \"\"\"\n",
    "        from PgenReader.read_list\n",
    "        \"\"\"\n",
    "        sample_idx = np.array(sample_idx) if sample_idx is not None else np.array(range(self.sample_ct))\n",
    "\n",
    "        buf = np.empty((len(variant_idxs), self.sample_ct), dtype=np.int32) # shape (variant_nums, sample_nums)\n",
    "\n",
    "        variant_idxs_array = np.array(variant_idxs, dtype=np.uint32)\n",
    "        self.pgen.read_list(variant_idxs = variant_idxs_array, \n",
    "                            geno_int_out = buf,\n",
    "                            allele_idx = allele_idx,\n",
    "                            )\n",
    "        return buf[:, sample_idx]\n",
    "\n",
    "    @overload\n",
    "    def extract(self, variant_idx:List[int], variant_ids=None, allele_idx=1, sample_idx=None)-> np.ndarray:...\n",
    "    @overload\n",
    "    def extract(self, variant_idx:int, variant_ids=None, allele_idx=1, sample_idx=None)-> np.ndarray:...\n",
    "    @overload\n",
    "    def extract(self, variant_ids:List[str], variant_idx=None, allele_idx=1, sample_idx=None)-> np.ndarray:...\n",
    "    @overload\n",
    "    def extract(self, variant_ids:str, variant_idx=None, allele_idx=1, sample_idx=None)-> np.ndarray:...\n",
    "    \n",
    "    def extract(self, variant_idx=None, variant_ids=None, allele_idx=1, sample_idx=None)->np.ndarray:\n",
    "        \"\"\"\n",
    "        TF version for extract data from pgen file \n",
    "        Usage:\n",
    "        pgfull = PgenReaderFull(pfile_path) # suffix of pgen,pvar and psam like plink2 usage do \n",
    "\n",
    "        # extract the first 3 variants, the second allele, and the first 3 samples\n",
    "        snp_idxs = [1,2,3]\n",
    "        geno, variants, samples = pgfull.extract(variant_idx=[1,2,3])\n",
    "\n",
    "        # or extract by variant_ids\n",
    "        geno, variants, samples = pgfull.extract(variant_ids=['rs1','rs2','rs3'])\n",
    "\n",
    "        # or only part of the samples\n",
    "        geno, variants, samples = pgfull.extract(variant_idx=[1,2,3], sample_idx=[1,2,3])\n",
    "\n",
    "        \"\"\"\n",
    "        if variant_idx is None and variant_ids is None:\n",
    "            raise ValueError('variant_idx and variant_ids cannot be None at the same time')\n",
    "        if variant_idx and variant_ids:\n",
    "            raise ValueError('variant_idx and variant_ids cannot be assigned at the same time')\n",
    "        if variant_idx:\n",
    "            if isinstance(variant_idx, int):\n",
    "                variant_idx = [variant_idx]\n",
    "\n",
    "            extracted_geno = self.read_list(variant_idx, allele_idx, sample_idx)\n",
    "            samples = self.get_sample_ids(sample_idx) if sample_idx else self.sample_list\n",
    "            variants = self.get_variant_ids(variant_idx)\n",
    "\n",
    "        if variant_ids:\n",
    "            if isinstance(variant_ids, str):\n",
    "                variant_ids = [variant_ids]\n",
    "\n",
    "            variant_idx = self.get_variant_idx(variant_ids)\n",
    "\n",
    "            extracted_geno = self.read_list(variant_idx, allele_idx, sample_idx)\n",
    "            samples = self.get_sample_ids(sample_idx) if sample_idx else self.sample_list\n",
    "            variants = variant_ids\n",
    "\n",
    "        # return pd.DataFrame(extracted_geno, index=variants, columns=samples)\n",
    "        return extracted_geno, variants, samples\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    def read_snp_list(self, snp_list, allele_idx=1, sample_idx=None)->np.ndarray:\n",
    "        pass \n",
    "\n",
    "\n",
    "    def read_dosage(self):\n",
    "        \"\"\"\n",
    "        soon modify from PgenReader.read_dosage\n",
    "        \"\"\"\n",
    "        raise NotImplementedError\n",
    "    def read_alleles(self):\n",
    "        \"\"\"\n",
    "        soon modify from PgenReader.read_alleles\n",
    "        \"\"\"\n",
    "        raise NotImplementedError\n",
    "    \n",
    "    def read_alleles_and_phasepresent(self):\n",
    "        \"\"\"\n",
    "        soon modify from PgenReader.read_alleles_and_phasepresent\n",
    "        \"\"\"\n",
    "        raise NotImplementedError\n",
    "    \n",
    "\n",
    "    \n",
    "    def __enter__(self):\n",
    "        return self\n",
    "    def __exit__(self):\n",
    "        self.close()\n",
    "\n",
    "    def close(self):\n",
    "        \"\"\"\n",
    "        from PgenReader.close\n",
    "        \"\"\"\n",
    "        self.pgen.close()\n",
    "        self.pvar.close()\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Union, List, overload\n",
    "# from \n",
    "\n",
    "class PgenReaderFull():\n",
    "    def __init__(self, pfile_path:str, pgen_path:str =None, pvar_path:str=None, sample_path:str=None ):\n",
    "\n",
    "        self.pgen_path:str = pgen_path if pgen_path is not None else pfile_path + '.pgen'\n",
    "        self.pvar_path:str = pvar_path if pvar_path is not None else pfile_path + '.pvar'\n",
    "        self.sample_path:str = sample_path if sample_path is not None else pfile_path + '.psam'\n",
    "\n",
    "        self.pvar:pg.PvarReader = pg.PvarReader(bytes(self.pvar_path, 'utf-8'))\n",
    "        self._init_var() # this may be too large to load\n",
    "        \n",
    "        self.sample:pd.DataFrame = pd.read_csv(self.sample_path, sep='\\t')\n",
    "        self._init_sample()\n",
    "\n",
    "        self.pgen:pg.PgenReader = pg.PgenReader(bytes(self.pgen_path, 'utf-8'), pvar=self.pvar, \n",
    "                                                sample_subset=np.arange(len(self.sample_list), dtype=np.uint32)\n",
    "                                                )\n",
    "\n",
    "\n",
    "        # basic information\n",
    "        self.sample_ct:int = self.get_raw_sample_ct() # number of samples as same as get_raw_sample_ct, but only call once\n",
    "        self.variant_ct:int = self.get_variant_ct() # number of variants as same as get_variant_ct, but only call once\n",
    "\n",
    "\n",
    "    @overload\n",
    "    def get_variant_ids(self, variant_idx:List[int])->List[str]:...\n",
    "    @overload\n",
    "    def get_variant_ids(self, variant_idx:int)->str:...\n",
    "\n",
    "    def get_variant_ids(self, variant_idx:Union[int, List[int]])->Union[str, List[str]]:\n",
    "        \"\"\"\n",
    "        from PvarReader.get_variant_id \n",
    "        Modified version for interface        \n",
    "        \"\"\"\n",
    "        if isinstance(variant_idx, int):\n",
    "            return self.pvar.get_variant_id(variant_idx).decode('utf-8')\n",
    "        else:\n",
    "            return [self.get_variant_ids(i) for i in variant_idx]\n",
    "        \n",
    "    @overload\n",
    "    def get_variant_idx(self, variant_id:str)->int:...\n",
    "    @overload\n",
    "    def get_variant_idx(self, variant_id:List[str])->List[int]:...\n",
    "    def get_variant_idx(self, variant_id:Union[str, List[str]])->Union[int, List[int]]:\n",
    "        \"\"\"\n",
    "        Modified version for interface\n",
    "        \"\"\"\n",
    "        if isinstance(variant_id, str):\n",
    "            return self.var_list.index(variant_id)\n",
    "        else:\n",
    "            return [self.get_variant_idx(i) for i in variant_id]\n",
    "    def get_sample_ids(self, sample_idx:Union[int, List[int]])->Union[str, List[str]]:\n",
    "        \"\"\"\n",
    "        TF version for get sample ids\n",
    "        \"\"\"\n",
    "        if isinstance(sample_idx, int):\n",
    "            return self.sample_list[sample_idx]\n",
    "        else:\n",
    "            return [self.get_sample_ids(i) for i in sample_idx]\n",
    "\n",
    "\n",
    "    def _init_sample(self):\n",
    "        \"\"\"\n",
    "        load from psam and to list[[FID, IID]]\n",
    "        sample_list = [[FID, IID], ...]\n",
    "        \"\"\"\n",
    "        self.sample_list = pd.read_csv(self.sample_path, sep='\\t').iloc[:, :2].values.tolist()\n",
    "    def _init_var(self):\n",
    "        \"\"\"\n",
    "        load from pvar and to list[variant_id]\n",
    "        \"\"\"\n",
    "        self.var_list = pd.read_csv(self.pvar_path, sep='\\t').iloc[:, 2].tolist()\n",
    "\n",
    "    def get_raw_sample_ct(self) -> int:\n",
    "        \"\"\"\n",
    "        from PgenReader.get_raw_sample_ct\n",
    "        \"\"\"\n",
    "        return self.pgen.get_raw_sample_ct()\n",
    "\n",
    "    def get_variant_ct(self) -> int:\n",
    "        \"\"\"\n",
    "        from PgenReader.get_variant_ct\n",
    "        \"\"\"\n",
    "        return self.pgen.get_variant_ct()\n",
    "    \n",
    "    def hardcall_phase_present(self) -> bool:\n",
    "        \"\"\"\n",
    "        from PgenReader.hardcall_phase_present\n",
    "        \"\"\"\n",
    "        return self.pgen.hardcall_phase_present()\n",
    "    def read(self, variant_idx, sample_idx=None, allele_idx=1)->np.ndarray:\n",
    "        \"\"\"\n",
    "        from PgenReader.read with modified \n",
    "\n",
    "        Parameters\n",
    "            - variant_idx:int32, index of the variant to read\n",
    "            - allele_idx (int, optional): 等位基因索引，默认为 1。当 allele_idx 为 1 时，返回的基因型信息是基于ALT等位基因的计数；当 allele_idx 为 0 时，返回的基因型信息是基于REF等位基因的计数。\n",
    "\n",
    "        \"\"\"\n",
    "        \n",
    "        sample_idx = np.array(sample_idx) if sample_idx is not None else np.array(range(self.sample_ct))\n",
    "\n",
    "        # use buf to store the result; the author of pgenlib think this is faster \n",
    "        buf = np.empty(self.sample_ct, dtype=np.int8)\n",
    "        self.pgen.read(variant_idx = variant_idx, \n",
    "                       geno_int_out = buf,\n",
    "                       allele_idx = allele_idx,\n",
    "                       \n",
    "                       )\n",
    "        return buf[sample_idx]\n",
    "    def read_range(self, variant_idx_start, variant_idx_end, allele_idx=1, sample_idx=None)->np.ndarray:\n",
    "        \"\"\"\n",
    "        from PgenReader.read_range\n",
    "        Parameters\n",
    "            - variant_idx_start: int32, index of the first variant to read\n",
    "            - variant_idx_end: int32, index of the last variant to read\n",
    "            - allele_idx: int32, index of the allele to read, defualt 1 for ALT count, while 0 for REF count\n",
    "            - sample_idx: int32, index of the sample to read\n",
    "        \"\"\"\n",
    "        sample_idx = np.array(sample_idx) if sample_idx is not None else np.array(range(self.sample_ct))\n",
    "\n",
    "        buf = np.empty((self.sample_ct, variant_idx_end-variant_idx_start), dtype=np.int8)\n",
    "        self.pgen.read_range(variant_idx_start = variant_idx_start, \n",
    "                             variant_idx_end = variant_idx_end,\n",
    "                             geno_int_out = buf,\n",
    "                             allele_idx = allele_idx,\n",
    "                             )\n",
    "        \n",
    "        return buf[:, sample_idx]\n",
    "    def read_list(self, variant_idxs:list,  allele_idx=1, sample_idx=None) -> np.ndarray:\n",
    "        \"\"\"\n",
    "        from PgenReader.read_list\n",
    "        \"\"\"\n",
    "        sample_idx = np.array(sample_idx) if sample_idx is not None else np.array(range(self.sample_ct))\n",
    "\n",
    "        buf = np.empty((len(variant_idxs), self.sample_ct), dtype=np.int32) # shape (variant_nums, sample_nums)\n",
    "\n",
    "        variant_idxs_array = np.array(variant_idxs, dtype=np.uint32)\n",
    "        self.pgen.read_list(variant_idxs = variant_idxs_array, \n",
    "                            geno_int_out = buf,\n",
    "                            allele_idx = allele_idx,\n",
    "                            )\n",
    "        return buf[:, sample_idx]\n",
    "\n",
    "    @overload\n",
    "    def extract(self, variant_idx:List[int], variant_ids=None, allele_idx=1, sample_idx=None)-> np.ndarray:...\n",
    "    @overload\n",
    "    def extract(self, variant_idx:int, variant_ids=None, allele_idx=1, sample_idx=None)-> np.ndarray:...\n",
    "    @overload\n",
    "    def extract(self, variant_ids:List[str], variant_idx=None, allele_idx=1, sample_idx=None)-> np.ndarray:...\n",
    "    @overload\n",
    "    def extract(self, variant_ids:str, variant_idx=None, allele_idx=1, sample_idx=None)-> np.ndarray:...\n",
    "    \n",
    "    def extract(self, variant_idx=None, variant_ids=None, allele_idx=1, sample_idx=None)->np.ndarray:\n",
    "        \"\"\"\n",
    "        TF version for extract data from pgen file \n",
    "        Usage:\n",
    "        pgfull = PgenReaderFull(pfile_path) # suffix of pgen,pvar and psam like plink2 usage do \n",
    "\n",
    "        # extract the first 3 variants, the second allele, and the first 3 samples\n",
    "        snp_idxs = [1,2,3]\n",
    "        geno, variants, samples = pgfull.extract(variant_idx=[1,2,3])\n",
    "\n",
    "        # or extract by variant_ids\n",
    "        geno, variants, samples = pgfull.extract(variant_ids=['rs1','rs2','rs3'])\n",
    "\n",
    "        # or only part of the samples\n",
    "        geno, variants, samples = pgfull.extract(variant_idx=[1,2,3], sample_idx=[1,2,3])\n",
    "\n",
    "        \"\"\"\n",
    "        if variant_idx is None and variant_ids is None:\n",
    "            raise ValueError('variant_idx and variant_ids cannot be None at the same time')\n",
    "        if variant_idx and variant_ids:\n",
    "            raise ValueError('variant_idx and variant_ids cannot be assigned at the same time')\n",
    "        if variant_idx:\n",
    "            if isinstance(variant_idx, int):\n",
    "                variant_idx = [variant_idx]\n",
    "\n",
    "            extracted_geno = self.read_list(variant_idx, allele_idx, sample_idx)\n",
    "            samples = self.get_sample_ids(sample_idx) if sample_idx else self.sample_list\n",
    "            variants = self.get_variant_ids(variant_idx)\n",
    "\n",
    "        if variant_ids:\n",
    "            if isinstance(variant_ids, str):\n",
    "                variant_ids = [variant_ids]\n",
    "\n",
    "            variant_idx = self.get_variant_idx(variant_ids)\n",
    "\n",
    "            extracted_geno = self.read_list(variant_idx, allele_idx, sample_idx)\n",
    "            samples = self.get_sample_ids(sample_idx) if sample_idx else self.sample_list\n",
    "            variants = variant_ids\n",
    "\n",
    "        # return pd.DataFrame(extracted_geno, index=variants, columns=samples)\n",
    "        return extracted_geno, variants, samples\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    def read_snp_list(self, snp_list, allele_idx=1, sample_idx=None)->np.ndarray:\n",
    "        pass \n",
    "\n",
    "\n",
    "    def read_dosage(self):\n",
    "        \"\"\"\n",
    "        soon modify from PgenReader.read_dosage\n",
    "        \"\"\"\n",
    "        raise NotImplementedError\n",
    "    def read_alleles(self):\n",
    "        \"\"\"\n",
    "        soon modify from PgenReader.read_alleles\n",
    "        \"\"\"\n",
    "        raise NotImplementedError\n",
    "    \n",
    "    def read_alleles_and_phasepresent(self):\n",
    "        \"\"\"\n",
    "        soon modify from PgenReader.read_alleles_and_phasepresent\n",
    "        \"\"\"\n",
    "        raise NotImplementedError\n",
    "    \n",
    "\n",
    "    \n",
    "    def __enter__(self):\n",
    "        return self\n",
    "    def __exit__(self):\n",
    "        self.close()\n",
    "\n",
    "    def close(self):\n",
    "        \"\"\"\n",
    "        from PgenReader.close\n",
    "        \"\"\"\n",
    "        self.pgen.close()\n",
    "        self.pvar.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "work",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.1.undefined"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
